### PR TITLE
add rx-store

### PR DESCRIPTION
**Pull request by joshribakoff**
I added [Rx Store.](https://rx-store.github.io/rx-store/)

I don't get any tearing, but it warns me about exceeding a max count, however it's not obvious to me why this is the wrong behavior, the test harness is in fact calling the provided increment thunk twice. I'm not sure I understand the exact condition the test harness is trying to accomplish here, maybe framing it as a real world use case would help (or what am I missing?)

https://www.youtube.com/watch?v=itWbTZWIcVU&feature=youtu.be

### CODE REVIEW

**Comment by dai-shi** [comment]
Hey, I hope you enjoy our conversation on Twitter.
I'd admit that "exceeding count" is far from intuitive. Having been difficult explaining it.
You want to check the behavior first with the browser.
What I'd like to test, and what I'm supposing to see is, when "Pending..." is shown, we should not see any counts changing. Only when "Pending..." is gone, counts should be updated.

re-rxjs (react-rxjs) somehow passes it, and my assumption now is that my tool is not correctly detecting it. I'd like to revisit it later, and also want somebody to help.

tearing and state branching would be fairly hypothetical issue and behavior, I'd say. So, it wouldn't bother too much for app developers. For lib authors, they may or may not.

File: src/rx-store/index.js
Comment (dai-shi) [inline src/rx-store/index.js:30]: Would you check initialState and reducer? it conditionally updates count. (which again is not very intuitive, I had to add this, because some of tests bypassed without this.)

Comment (joshribakoff) [reply src/rx-store/index.js]: I passed in your reducer, flux action, and initial state & it no longer has the error

File: src/rx-store/index.js
Comment (dai-shi) [inline src/rx-store/index.js:18]: nice!

**Comment by joshribakoff** [comment]
Just saw the test runner, here's the results it gives. I'd have to look into why this happens, but if anyone wants to explain to me that would be great too :D

I still have plans to build an example app that uses transitions, not sure what `rx-store` is doing differently than other rxjs libs that pass all checks... It seems to be more or less the same approach.

Screen Shot 2020-08-30 at 9 58 39 PM

Took a look at `rxjs-hooks`, saw they have `asapScheduler`. Added that & now 1 test fails. Hopefully I have time to look into it sometime soon. For now I'm willing to accept `startTransition` may or may not work & merge this as-is, if you're cool with it.

Screen Shot 2020-08-30 at 10 09 03 PM


PS > very much enjoying the Twitter convos!

**Comment by dai-shi** [comment]
For me, with the mind what I would like to test, failing check 7 and check 9 is reasonable. As I don't know `asapScheduler` very well, I'm not sure if it's a proper fix or somehow bypassing. I'm not confident if I was able to implement check 9 as I wanted. (The purpose of check 9 is to distinguish the behavior of useSubscription and useMutableSource described in https://gist.github.com/bvaughn/054b82781bec875345bd85a5b1344698)

re-rxjs passing check 7 is just unexpected. I would like to fix it, it should fail from my understanding. (And that's the discussion on Twitter, I'm not describing well.)

File: yarn.lock
Comment (dai-shi) [inline yarn.lock:1233]: We'd like to test against react@experimental. Is there any reason you don't make this peer dependency? I don't thinks it's very important, though.

**Review by dai-shi — APPROVED** [review]
The code looks good. Let merge this in and run all tests on my end.

**Comment by dai-shi** [comment]
@joshribakoff I updated check 7 completely. Now, it should make a little more sense.

**Comment by joshribakoff** [comment]
@dai-shi awesome just dropping in to say thanks, and I noticed everyone is failing it now 😆  makes me feel less bad. I'll be working through the use cases for `useTransition` by looking at Relay modern & the React unit tests, and using this to drive my implementation, then I'll come back to look at your test harness & see if I can make it pass / if I have any suggestions on changing it. Thanks again for making this & the quick review!

**Comment by dai-shi** [comment]
Sounds nice. Look forward to talk to you later.
To be honest, I'm not very confident with the new test yet. It's certainly testing a specific behavior, but what I wanted to test for state branching can be different.

**Comment by joshribakoff** [comment]
Yeah I'm a little skeptical because I expect to see certain libs failing certain checks based on ways I was able to make them break locally. Once I wrap my head around it more, I'll let you know if I have any actual suggestions to improve it :)

**Comment by joshribakoff** [comment]
FYI I added useTransition to my example app, branching seems to work fine. https://youtu.be/yfszFZR2OoE I actually think I realize what you've been saying now on Twitter, in addition to what I just did in my demo, you want to also be able to update state within the state manager, and apply that update only inside the `startTransition` callback?

Eg. In my demo (https://github.com/rx-store/rx-store/pull/69), I use `getResource()` (exported by Rx Store), I then take the returned `resource` and set it into the state...

And I guess what you want to do, is update "state" within Rx Store at the point in time that React applies the transition?

IMO this React mental model with the branching won't be scalable for complex business logic, I most likely won't support this use case. The way I would propose doing this is to store that state inside your `resource`. The whole reason I'm making Rx Store is to provide more elegant solutions where [React's paradigms fall short](https://reactjs.org/docs/concurrent-mode-suspense.html#race-conditions-with-useeffect).

IMO I'm not worried about passing check 7 here, because it's not an actual use case I think would be useful to users of Rx Store (assuming I understand the use case here correctly). My demo is implementing idiomatic React suspense. If an end user wants to `next()` onto the Rx Store subjects inside the `startTransition()`, I would try to convince them they are "wrong". Instead, I would encourage them to think of that as the same as running a side effect in the render method. Nexting onto an rxJS is a side effect, startTransition callback can be thought of as render phase IMO, and side effects simply do not belong in render phase. There's better ways to do this in RxJS where you can set up the streams themselves to buffer their values until some other condition is met on some other stream, see for eg. [http://reactivex.io/documentation/operators/buffer.html](http://reactivex.io/documentation/operators/buffer.html)
that being said, I do see how NOT having this tradeoff could be desirable to some. There are probably other use cases where it could look like the RxJS based state managers are able to do things more elegantly than the React state based state managers, just didn't tradeoffs at the end of the day :)

File: yarn.lock
Comment (joshribakoff) [reply yarn.lock:1233]: It should be peer. I’ll fix this and ping once it’s fixed. Sorry

**Comment by dai-shi** [comment]
> you want to also be able to update state within the state manager, and apply that update only inside the startTransition callback?

Yes, only inside WIP branch. I mean I just want to reveal the difference.

> I most likely won't support this use case.

Agreed. My comment was it's probably not worth for rxjs.

> IMO I'm not worried about passing check 7 here

Little apologies here. I don't mean global state libs need to pass all tests. Even with tearing (check 2), it wouldn't be important for most use cases. I have this mind from the beginning, but made the repo title a bit aggressive, otherwise people wouldn't care. I mean, I'd like people know this and discuss about it.

I added check 7, because we discussed this behavior and the limitation in redux community.

I'm doing this mostly from technical interests, but tearing seems to be an important issue for React CM, and that's why they made `useMutableSource`.

> I would try to convince them they are "wrong".

Makes sense. I'm not very familiar with RxJS, but Redux will probably take the same path.
For external stores, this is the "right" approach, because it's the single source.
For other libraries, they may want to support state branching. Constate cares about it from the early days (from the author's tweets). I guess React-Query will encourage startTransition in the future (it's just a guess).

Probably, the bad news is we will see the ecosystem division with CM. One side is React centric, the other is how do I call, if it's Redux, Redux centric. I'm not sure where the community goes, CM is nice and technically important, but maybe only for some use cases.

I'm developing global state libraries for both sides. So, it's important to see check 7 work correctly, the goal is not to pass it, but correctly passes and fails.

BTW, most of my personal use cases are small apps, so CM (and tearing) isn't so important for me as an app developer (it is important as a lib author though). However, Suspense/useTransition is very much interesting to me, because it has potential to improve DX.

**Pull request merged**
